lib/rubocop/cop/style/empty_line_between_defs.rb

```diff
@@ -9,27 +9,39 @@ class EmptyLineBetweenDefs < Cop
         MSG = 'Use empty lines between defs.'
 
         def on_def(node)
-          if @prev_def_end && @prev_def_end < def_end(node) &&
-             nothing_or_comments_between_end_and_def?(node)
-            unless @prev_was_single_line && single_line_def?(node) &&
-                   cop_config['AllowAdjacentOneLineDefs']
-              add_offense(node, :keyword)
-            end
-          end
+          nodes = [prev_node(node), node]
 
-          @prev_def_end = def_end(node)
-          @prev_was_single_line = single_line_def?(node)
+          return unless nodes.all?(&method(:def_node?))
+          return if blank_lines_between?(*nodes)
+          return if nodes.all?(&method(:single_line_def?)) &&
+                    cop_config['AllowAdjacentOneLineDefs']
+
+          add_offense(node, :keyword)
         end
 
         private
 
-        def nothing_or_comments_between_end_and_def?(node)
-          distance = def_start(node) - @prev_def_end
-          return true if distance == 1
+        def def_node?(node)
+          node && node.def_type?
+        end
+
+        def blank_lines_between?(first_def_node, second_def_node)
+          blank_re = /\A\s*\z/
+
+          lines_between_defs(first_def_node, second_def_node)
+            .any? { |line| line =~ blank_re }
+        end
+
+        def prev_node(node)
+          return nil unless node.parent && node.sibling_index > 0
+
+          node.parent.children[node.sibling_index - 1]
+        end
+
+        def lines_between_defs(first_def_node, second_def_node)
+          line_range = def_end(first_def_node)..(def_start(second_def_node) - 2)
```

Comment:
I'm not sure if this is the best way to get lines between `def` nodes, I especially don't like that `- 2` at the end. Can anyone suggest a better way to implement this?